examples/pipeline_wavernn/datasets.py

```diff
@@ -0,0 +1,125 @@
+import os
+import random
+
+import torch
+import torchaudio
+from torchaudio.datasets import LJSPEECH
+
+from transform import linear_to_mel
+from utils import label_to_waveform, mulaw_encode, specgram_normalize, waveform_to_label
+
+
+class MapMemoryCache(torch.utils.data.Dataset):
+    r"""Wrap a dataset so that, whenever a new item is returned, it is saved to memory.
+    """
+
+    def __init__(self, dataset):
+        self.dataset = dataset
+        self._cache = [None] * len(dataset)
+
+    def __getitem__(self, n):
+        if self._cache[n] is not None:
+            return self._cache[n]
+
+        item = self.dataset[n]
+        self._cache[n] = item
+
+        return item
+
+    def __len__(self):
+        return len(self.dataset)
+
+
+class ProcessedLJSPEECH(LJSPEECH):
+    def __init__(self, files, transforms, args):
+
+        self.transforms = transforms
+        self.files = files
+        self.args = args
+
+    def __getitem__(self, index):
+
+        file = self.files[index]
+        args = self.args
+        n_fft = 2048
+        waveform, sample_rate = torchaudio.load(file)
+        specgram = self.transforms(waveform)
+        # Will be replaced by torchaudio as described in https://github.com/pytorch/audio/pull/593
+        specgram = linear_to_mel(specgram, sample_rate, n_fft, args.n_freq, args.f_min)
+        specgram = specgram_normalize(specgram, args.min_level_db)
+        waveform = waveform.squeeze(0)
+
+        if args.mode == "waveform":
+            waveform = (
+                mulaw_encode(waveform, 2 ** args.n_bits)
+                if args.mulaw
+                else waveform_to_label(waveform, args.n_bits)
+            )
+
+        return waveform, specgram
+
+    def __len__(self):
+        return len(self.files)
+
+
+def datasets_ljspeech(args, transforms):
+
+    root = args.file_path
+    wavefiles = [os.path.join(root, file) for file in os.listdir(root)]
```

Comment:
We should organize the code in a way that doesn't require walking the files, but only use the utilities already available in torchaudio. Maybe we need a new one for this :) Thoughts?

Reply:
Now I use the `_walker` from LJSPEECH from torchaudio to get the files instead of walking the files.

Reply:
We shouldn't recommend that users access `_walker` since it is a private function as noted by the underscore. How about using a dictionary to map random subset of indices to original indices, and wrapping this into a class? Anyone else has other suggestions? @cpuhrsch @zhangguanheng66